inference/core/workflows/execution_engine/entities/base.py

```diff
@@ -157,20 +153,19 @@ def __iter__(self) -> Iterator[B]:
         yield from self._content
 
     def remove_by_indices(self, indices_to_remove: Set[tuple]) -> "Batch":
-        content, new_indices = [], []
-        for index, element in self.iter_with_indices():
-            if index in indices_to_remove:
-                continue
-            content.append(element)
-            new_indices.append(index)
-        return Batch(
-            content=content,
-            indices=new_indices,
-        )
+        filtered_content = [
+            element
+            for index, element in zip(self._indices, self._content)
+            if index not in indices_to_remove
+        ]
+        filtered_indices = [
+            index for index in self._indices if index not in indices_to_remove
+        ]
+
+        return Batch(content=filtered_content, indices=filtered_indices)
 
     def iter_with_indices(self) -> Iterator[Tuple[Tuple[int, ...], B]]:
-        for index, element in zip(self._indices, self._content):
-            yield index, element
+        return zip(self._indices, self._content)
```

Comment:
a couple of options here - I think that the `iter_with_indices` function can be updated to zip directly. It removes overhead. Then we can also make the line 158 in the new code to use the new `iter_with_indices` function